### ISSUE TITLE
Type "Shape[int]" is not assignable to declared type "Shape[int]"

### ISSUE THREAD

**Issue by jorenham** [microsoft/pyright#10466]
```py
class Shape[*Ts](tuple[*Ts]): ...

reveal_type(Shape([42]))
s1: Shape[int] = Shape([42])
```

reports

- ```
  Type of "Shape([42])" is "Shape[int]"
  ```
- ```
  Type "Shape[int]" is not assignable to declared type "Shape[int]"
    "Untitled.Shape" is not assignable to "Untitled.Shape"
      Type parameter "Ts@Shape" is invariant, but "int" is not the same as "*tuple[int]"  (reportAssignmentType)
  ```

which is... odd

[playground](https://pyright-play.net/?pyrightVersion=1.1.400&pythonVersion=3.14&strict=true&code=MYGwhgzhAEDKAWYAOBTA2gKgCoQLoAoAXAVyRHWzwEoAuaAOkYCgmAnFANxTBAH1CAnqnwJkKfGgAsAJlxUq0RUoDEcRKjQBLAHaFcTCAEY6ojTr3QAvGrESZcxaoCiddkgD2rQgEEomgObaALYoulhCKExAA)

**Comment by iwakitakuma33** [comment]
@jorenham 
It is already recognized as int [here](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L22726).


```
class Shape[*Ts](tuple[*Ts]): ...

s3_int: Shape[int] = Shape([(1)]) # error -> not correct
s3_tuple: Shape[int] = Shape([(1,)]) # OK

s_int: int = (1)
s_tuple: tuple[int] = (1,)

```

**Comment by jorenham** [comment]
so at runtime

```pycon
>>> class Shape[*Ts](tuple[*Ts]): ...
... 
>>> Shape([(42,)])[0]
(42,)
>>> 
```

yet `reveal_type(Shape([(42,)])[0])` is reported as `int`.

Sounds like it might indeed be related 🤔

**Comment by iwakitakuma33** [comment]
@jorenham 
That's right. I'll look into it.

```
from typing import reveal_type


class Shape[*Ts](tuple[*Ts]): ...


s1: Shape[int] = Shape([(1)])
# error: Type "Shape[int]" is not assignable to declared type "Shape[int]"
# "pyright_test.Shape" is not assignable to "pyright_test.Shape"
# Type parameter "Ts@Shape" is invariant, but "int" is not the same as "*tuple[int]" (reportAssignmentType)
reveal_type(s1[0])
# python 3.13 Runtime type is 'int'
# pyright Type of "s1[0]" is "int"
print(type(s1[0]), s1[0], s1)
# python 3.13 <class 'int'> 1 (1,)

s2: Shape[int] = Shape([(1,)])
# OK
reveal_type(s2[0])
# python 3.13 Runtime type is 'tuple'
# pyright Type of "s2[0]" is "int"
print(type(s2[0]), s2[0], s2)
# python 3.13 <class 'tuple'> (1,) ((1,),)

s3: Shape[int, int] = Shape(
    [
        (
            1,
            1,
        )
    ]
)
# OK
reveal_type(s3[0])
# python 3.13 Runtime type is 'tuple'
# pyright Type of "s3[0]" is "int"
print(type(s3[0]), s3[0], s3)
# python 3.13 <class 'tuple'> (1, 1) ((1, 1),)

s4: Shape[int, str] = Shape(
    [
        (
            1,
            "1",
        )
    ]
)
# OK
reveal_type(s4[0])
# python 3.13 Runtime type is 'tuple'
# pyright Type of "s4[0]" is "int"
print(type(s4[0]), s4[0], s4)
# python 3.13 <class 'tuple'> (1, '1') ((1, '1'),)


s5: Shape[int, str] = Shape(
    [
        (
            1,
            "1",
        ),
        (
            1,
            "1",
        ),
    ]
)
# OK
reveal_type(s5[0])
# python 3.13 Runtime type is 'tuple'
# pyright Type of "s5[0]" is "int"
print(type(s5[0]), s5[0], s5)
# python 3.13 <class 'tuple'> (1, '1') ((1, '1'), (1, '1'))


```
